RenderStatePackager/src/RenderStatePackager.cpp

```diff
@@ -33,217 +33,314 @@ namespace Diligent
-        }
+        m_pThreadPool->WaitForAllTasks();
 
-        for (Uint32 PipelineID = 0; PipelineID < ParserInfo.ComputePipelineStateCount; PipelineID++)
-        {
-            auto                           pResourceDescRSN = pNotationParser->GetComputePipelineStateByIndex(PipelineID);
-            ComputePipelineStateCreateInfo ResourceDesc     = {};
-
-            UnpackPipelineStateCreateInfo(*pResourceDescRSN, ResourceDesc);
-            ResourceDesc.pCS = FindShader(pResourceDescRSN->pCSName);
+        for (auto& Exception : Exceptions)
+            if (Exception)
+                std::rethrow_exception(Exception);
+    }
+    catch (...)
+    {
+        return false;
+    }
+    return true;
+}
 
-            PipelineStateArchiveInfo ArchiveInfo = {};
-            ArchiveInfo.DeviceFlags              = m_DeviceBits;
+bool RenderStatePackager::Execute(RefCntAutoPtr<IArchiver> pArchive)
+{
+    DEV_CHECK_ERR(pArchive != nullptr, "pArchive must not be null");
 
-            if (!pArchive->AddComputePipelineState(ResourceDesc, ArchiveInfo))
-                LOG_ERROR_AND_THROW("Failed to archive compute pipeline '", ResourceDesc.PSODesc.Name, "'.");
-        }
+    if (pArchive == nullptr)
+        return false;
 
-        for (Uint32 PipelineID = 0; PipelineID < ParserInfo.TilePipelineStateCount; PipelineID++)
+    try
+    {
+        std::vector<std::exception_ptr>        Exceptions;
+        std::vector<RefCntAutoPtr<IAsyncTask>> Tasks;
+        for (auto const& pNotationParser : m_RSNParsers)
         {
-            auto pResourceDescRSN = pNotationParser->GetTilePipelineStateByIndex(PipelineID);
+            auto const& ParserInfo = pNotationParser->GetInfo();
 
-            TilePipelineStateCreateInfo ResourceDesc = {};
-            UnpackPipelineStateCreateInfo(*pResourceDescRSN, ResourceDesc);
+            for (Uint32 ShaderID = 0; ShaderID < ParserInfo.ShaderCount; ShaderID++)
+            {
+                auto pTask = EnqueueAsyncWork(m_pThreadPool, [ShaderID, pNotationParser, this, &Exceptions](Uint32 ThreadId) {
+                    try
+                    {
+                        ShaderCreateInfo ShaderCI           = *pNotationParser->GetShaderByIndex(ShaderID);
+                        ShaderCI.pShaderSourceStreamFactory = m_pStreamFactory;
+
+                        RefCntAutoPtr<IShader> pShader;
+                        m_pDevice->CreateShader(ShaderCI, m_DeviceBits, &pShader);
+                        if (!pShader)
+                            LOG_ERROR_AND_THROW("Failed to create shader from file '", ShaderCI.FilePath, "'.");
+
+                        std::unique_lock<std::mutex> Lock{m_LockShaders};
+                        m_Shaders.emplace(ShaderCI.Desc.Name, pShader);
+                    }
+                    catch (...)
+                    {
+                        Exceptions.push_back(std::current_exception());
```

Comment:
Exceptions used without mutex - resize to m_RSNParsers.size() to avoid the problem